File: Java/1365_How_Many_Numbers_Are_Smaller_Than_the_Current_Number.java
class Solution {
    public int[] smallerNumbersThanCurrent(int[] nums) {
        int[] numsCopy = Arrays.copyOf(nums, nums.length);
        Arrays.sort(numsCopy);
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < numsCopy.length; i++){
            if(map.get(numsCopy[i]) == null){
                map.put(numsCopy[i], i);
            }
        }
        
        int[] result = new int[nums.length];
        for(int i = 0; i < nums.length; i++){
            result[i] = map.get(nums[i]);
        }
        return result;
        /*
        int[] result = new int[nums.length];
        for(int i = 0; i < nums.length; i++){
            int total = 0;
            for(int j = 0; j < nums.length; j++){
                if(nums[i] > nums[j]){
                    total++;
                }
            }
            result[i] = total;
        }
        return result;
        */
    }
}

File: CS/1365_How_Many_Numbers_Are_Smaller_Than_the_Current_Number.cs
public class Solution {
    public int[] SmallerNumbersThanCurrent(int[] nums) {
        int[] copy = new int[nums.Length];
        Array.Copy(nums, 0, copy, 0, nums.Length);
        Array.Sort(copy);
        Dictionary<int, int> dic = new Dictionary<int, int>();
        for(int i = 0; i < copy.Length; i++){
            if(!dic.ContainsKey(copy[i])){
                dic.Add(copy[i], i);
            }
        }
        
        int[] result = new int[nums.Length];
        for(int i = 0; i < nums.Length; i++){
            result[i] = dic[nums[i]];
        }
        return result;
        /*
        int[] result = new int[nums.Length];
        for(int i = 0; i < nums.Length; i++){
            for(int j = 0; j< nums.Length; j++){
                if(nums[j] < nums[i]){
                    result[i]++;
                }
            }
        }
        return result;
        */
    }
}
